python/cudf/dataframe/string.py

```diff
@@ -590,6 +592,64 @@ def to_array(self, fillna=None):
 
         return self.to_arrow().to_pandas()
 
+    def serialize(self, serialize):
```

Comment:
@kkraus14 @davidwendt This is my rough attempt to serialize/deserialize `StringColumn` data.  Seems to "work" fine, but please let me know if there is a more-efficient solution :)

Reply:
We should be able to do this entirely on the device using a combination of rmm and our `cudf.utils.cudautils`. If you follow the logic in the `to_arrow` method and replace the numpy calls with `cudautils` calls, we should be able to just pass an extra parameter to the `to_offsets` call of `bdevmem=True` to instruct it to use the device memory.

Reply:
Okay, that makes sense - Thanks!

Reply:
@kkraus14 Tahnks again for the pointers!  I am having some trouble understanding the correct way to use `to_offsets` with `DeviceNDArray` objects for the values and offsets.  Following the `to_arrow` code, I started with something like this:

```p
sbuf = cudautils.zeros(self._data.byte_count(), dtype='int8')
obuf = cudautils.zeros(len(self._data) + 1, dtype='int32')
mask_size = utils.calc_chunk_size(len(self._data), utils.mask_bitsize)
nbuf = cudautils.zeros(mask_size, dtype='int8')

self.str().to_offsets(sbuf, obuf, nbuf=nbuf, bdevmem=True)
```
Which throws an error:

`SystemError: <built-in function n_create_offsets> returned a result with an error set`

Reply:
> @kkraus14 Thanks again for the pointers! I am having some trouble understanding the correct way to use `to_offsets` with `DeviceNDArray` objects for the values and offsets. Following the `to_arrow` code, I started with something like this:
> 
> ```abl
> sbuf = cudautils.zeros(self._data.byte_count(), dtype='int8')
> obuf = cudautils.zeros(len(self._data) + 1, dtype='int32')
> mask_size = utils.calc_chunk_size(len(self._data), utils.mask_bitsize)
> nbuf = cudautils.zeros(mask_size, dtype='int8')
> 
> self.str().to_offsets(sbuf, obuf, nbuf=nbuf, bdevmem=True)
> ```
> 
> Which throws an error:
> 
> `SystemError: <built-in function n_create_offsets> returned a result with an error set`

Instead of `self.str().to_offsets(...)` you should use `self.data.to_offsets(...)` to directly call the function against the nvstrings object. I'm guessing what's happening is when going through the StringAccessor it's hitting our `__getattr__` shenanigans to try to shepherd calls to nvstrings but convert the result back to a cudf StringColumn.

Additionally, instead of using `cudautils.zeros(...)` you should be able to just use `cudf.rmm.device_array(...)` since you don't actually need the memory zeroed as far as I'm aware.

Reply:
Hmm, looks like this may be an issue within NVStrings itself, cc @davidwendt any insight here?

Reply:
You can get the pointer of an `rmm.device_array(...)` object using `cudf.bindings.cudf_cpp.get_ctype_ptr` which will return it as a Python integer which is what the function will expect.

Reply:
Sorry @kkraus14 - Still having issues with this.

Your comment on Slack:
>We want `bdevmem=True` and then we need to get the pointers to the `rmm.device_array` objects and pass those as the parameters to the `to_offsets(...)` function

What is the correct way to get the pointer?  `device_ctypes_pointer`?

Reply:
`cudf.bindings.cudf_cpp.get_ctype_ptr` of an `rmm.device_array(...)` it should return you the pointer as an integer

Reply:
@kkraus14  Sorry for bugging you so much!  The following is working fine for the *serialize* phase:
```p
sbuf = rmm.device_array(len(self._data) + 1, dtype='int8')
obuf = rmm.device_array(len(self._data) + 1, dtype='int32')
mask_size = utils.calc_chunk_size(len(self._data), utils.mask_bitsize)
nbuf = rmm.device_array(mask_size, dtype='int8')
self.data.to_offsets(get_ctype_ptr(sbuf), get_ctype_ptr(obuf), nbuf=get_ctype_ptr(nbuf), bdevmem=True)
```
However, I get a segmentation fault during the *deserialize* phase when I use the same (`get_ctype_ptr`) approach to call `nvstrings.from_offsets`.  Note that the call works fine when using `*.copy_to_host()` on the arguments instead.

Reply:
Looks like it doesn't support device memory in the `from_offsets` call. @davidwendt would this be difficult to add?

Reply:
Issue filed here: https://github.com/rapidsai/custrings/issues/340

Reply:
thanks!

Reply:
PR is merged in custrings to add device memory support for `from_offsets`, if you update nvstrings and libnvstrings it should work now. Thanks @davidwendt!